scripts/hca-utils/hca_utils/utils.py

```diff
@@ -0,0 +1,332 @@
+
+        dataset_id = self._hit_jade("enumerateDatasets", handle_response, query={"filter": self.dataset})
+        return dataset_id
+
+    def submit_soft_delete(self, target_table: str, target_path: str) -> str:
+        """
+        Submit a soft delete request.
+        :param target_table: The table to apply soft deletion to.
+        :param target_path: The gs-path of the csv that contains the row ids to soft delete.
+        :return: The job id of the soft delete job.
+        """
+        dataset_id = self.get_dataset_id()
+
+        body = {
+            "deleteType": "soft",
+            "specType": "gcsFile",
+            "tables": [
+                {
+                    "gcsFileSpec": {
+                        "fileType": "csv",
+                        "path": target_path
+                    },
+                    "tableName": target_table
+                }
+            ]
+        }
+
+        def handle_response(response):
+            return response.json()["id"]
+
+        job_id = self._hit_jade("applyDatasetDataDeletion", handle_response, body=body, params={"id": dataset_id})
+
+        return job_id
+
+    # dataset-level checking and soft deleting
+    def check_for_duplicates(self):
+        """
+        Check and print the number of duplicates for each table in the dataset.
+        :return:
+        """
+        self._process_rows(self.get_all_table_names, self.get_duplicates, soft_delete=False)
+
+    def check_for_null_file_refs(self):
+        """
+        Check and print the number of null file references for each table in the dataset that has a `file_id` column.
+        :return:
+        """
+        self._process_rows(self.get_file_table_names, self.get_null_filerefs, soft_delete=False)
+
+    def check_for_all(self):
+        """
+        Check and print the number of duplicates and null file references in all tables in the dataset.
+        :return:
+        """
+        self.check_for_duplicates()
+        self.check_for_null_file_refs()
+
+    def remove_duplicates(self):
+        """
+        Check and print the number of duplicates for each table in the dataset, then soft delete the problematic rows.
+        :return:
+        """
+        self._process_rows(self.get_all_table_names, self.get_duplicates)
+
+    def remove_null_file_refs(self):
+        """
+        Check and print the number of null file references for each table in the dataset, then soft delete the
+        problematic rows.
+        :return:
+        """
+        self._process_rows(self.get_file_table_names, self.get_null_filerefs)
+
+    def remove_all(self):
+        """
+        Check and print the number of duplicates and null file references for each table in the dataset, then soft
+        delete the problematic rows.
+        :return:
+        """
+        self.remove_duplicates()
+        self.remove_null_file_refs()
+
+    def _process_rows(self, get_table_names, get_rids, soft_delete: bool = True):
```

Comment:
This runs synchronously; if we care enough, I can try to make it async, but I figured it isn't really worth it for this

Reply:
worth making this async if it takes eternity to run and is mostly waiting for external services, but otherwise async is currently extremely fiddly in python

Reply:
I'm 👎 on async until it's absolutely necessary from a perf perspective, especially in a utility script like this.

Reply:
Yeah this all runs fast enough that I don't think we should bother